# Ejercicio 5
Alumna: Blanca Cano Camarero  
Correo: blancacano@correo.ugr.es

Dado un número $n$

In [1]:
n = 55883687
factor(n)

7393 * 7559

### Apartado primero 

Factoriza $n$ aplicando el método de la $\rho$ de Polard. ¿Cuántas iteraciones necesitas?
Sea $p1$ el mayor de sus factores primos y $p2$ el siguiente primo.

In [2]:
# Código auxiliar de prácticas ateriores
def Verbose_Rho_de_Polard(n:Integer, f) -> Integer:
    '''Devuelve un factor de p
    o n si es un fracaso
    '''
    mf = lambda x: f(x)% n
    #print(f"Cálculo de Rho de Polar para n={n}")
    x = 1
    y = 1
    d = 1
    paso = 0
    #print(f"Paso {paso}   x={x}    y={y}")
    while d == 1:
        paso += 1
        x = mf(x)
        y = mf(mf(y))
        d = gcd(abs(x-y),n)
        #print(f"Paso {paso}    x={x}    y={y}    mcd={d}")
    return d
f = lambda x : x^2 +1 

## Versión de sacar todos los factores de manera manual 
def SacaFactores(numero_factores, n, f ):
    factores = []
    m = n
    hay_factores = True 
    while numero_factores >= 0 and hay_factores:
        numero_factores -= 1
        d = Verbose_Rho_de_Polard(m,f)
        if d != m:
            factores.append(d)
            m = m // d
            print(f"Se tiene que n = multiplicar los elementos de {factores} y {m}\n")
        else:
            factores.append(m)
            print('='*60)
            hay_factores = False
    return factores

factores = SacaFactores(2, n,f )
print(f"Los factores de {n-1} son {factores}")

def Test_Lucas_Lehmer(n, bases):
    m = n-1
    for a in bases:
        # cumple condición 1
        if mod(a^m, n) == 1:
            condicion_dos = True
            for i in range (2,m):
                # si es divisor 
                if m % i == 0:
                    if (int(a**(m/i))% n) == 1:
                        condicion_dos =  False
                        break
            if condicion_dos == True:
                return a
    return 0

for p in factores:
    resultado = Test_Lucas_Lehmer(p,[2,3,5,7,11,13])
    mensaje = (
        f"p={p} es primo por el test Lucas-Lehmer con a = {resultado}",
        "No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = {p}"
    )[resultado == 0] 
    print(mensaje)

Se tiene que n = multiplicar los elementos de [7559] y 7393

Los factores de 55883686 son [7559, 7393]
p=7559 es primo por el test Lucas-Lehmer con a = 13
p=7393 es primo por el test Lucas-Lehmer con a = 5


A la vista de los resultado, tenemos que 
$p1 = 7559$ y $p2=7393$
### Apartado segundo   

Calcula las partes enteras de $\sqrt{p1}$ y $\sqrt{p2}$ con el algoritmo entero.

In [10]:
def VerboseRaizCuadradoEntero(n):
    # Incicializamos a 
    a = n
    if is_even(n):
        a //= 2
    else:
        a = (a+1)//2

    
    menor = True 
    i = 1
    while menor:
        numerador = a**2 + n
        cociente =  numerador//(2*a)
        resto = numerador % (2*a)
        print(f"i:{i}  a_i:{a}  a^2 +n: {numerador} = {cociente} * {2*a} + {resto}")
        i += 1
        if cociente < a:
            a = cociente
        else:
            menor = False 
    print( "\n",f"La parte entera de  sqr({n}) es{a}", '\n','='*50,'\n')   
    return a 
for p in [15,101]:#factores:
    VerboseRaizCuadradoEntero(p)

    

i:1  a_i:8  a^2 +n: 79 = 4 * 16 + 15
i:2  a_i:4  a^2 +n: 31 = 3 * 8 + 7
i:3  a_i:3  a^2 +n: 24 = 4 * 6 + 0

 La parte entera de  sqr(15) es3 

i:1  a_i:51  a^2 +n: 2702 = 26 * 102 + 50
i:2  a_i:26  a^2 +n: 777 = 14 * 52 + 49
i:3  a_i:14  a^2 +n: 297 = 10 * 28 + 17
i:4  a_i:10  a^2 +n: 201 = 10 * 20 + 1

 La parte entera de  sqr(101) es10 

